nepalipatro@deeppradhan/files/nepalipatro@deeppradhan/desklet.js

```diff
@@ -0,0 +1,364 @@
+		let fontFamily;
+		if (this.fontFamily !== "default") {
+			if (this.fontFamily === "custom-font-family" && this.customFontFamily !== "")
+				fontFamily = this.customFontFamily;
+			else
+				fontFamily = this.fontFamily;
+		}
+		if (typeof (this.boxLayoutCalendar) !== "undefined")
+			this.boxLayoutCalendar.remove_all_children();
+		this.boxLayoutCalendar = new St.BoxLayout({vertical: this.layout !== "horizontal"});
+		this.boxLayoutCalendar.style = "background-color: " + (this.colourBackground.replace(")", ","
+				+ (1 - this.transparency / 100) + ")")).replace('rgb', 'rgba')
+				+ "; border-radius: " + (this.fontSize / 3 * 2) + "pt; color: " + this.colourText + ";"
+				+ (typeof(fontFamily) !== "undefined" ? " font-family: '" + fontFamily + "';" : "")
+				+ " font-size: " + this.fontSize + "pt; padding: " + (this.fontSize / 3 * 2) + "pt; text-shadow: 1px 1px 2px #000;";
+		if (this.panels === "both" || this.panels === "today")
+			this.boxLayoutCalendar.add_actor(this.buttonToday);
+		if (this.panels === "both" || this.panels === "month")
+			this.boxLayoutCalendar.add_actor(this.tableMonth);
+		if (this.panels === "both")
+			this.buttonToday.style = "margin-" + (this.layout === "horizontal" ? "right" : "bottom")
+					+ ": " + (this.fontSize / 2) + "pt;";
+		
+		this.setContent(this.boxLayoutCalendar);
+
+		this.updateValues();
+	},
+
+	/* Method to update the Desklet values*/
+	updateValues: function() {
+
+		let now = new Date();
+		// Convert to Nepali TimeZone if specified
+		if (this.timeZone === "nepali")
+			now = NepaliPatro.dateWithOffset(now, 5.75);
+
+		if (this.lastUpdate.fullYear !== now.getFullYear() || this.lastUpdate.month !== now.getMonth() || this.lastUpdate.date !== now.getDate()) {
+			this.updateCalendar();
+			return;
+		}
+
+		let bsNow = new BikramSambat(now);
+		bsNow.language = this.language;
+		bsNow.monthTextShort = this.shortMonthName;
+
+		//////// Today Panel ////////
+		this.labelDay.set_text(this.showWeekday !== "full" ?
+				BikramSambat.prototype.weekdayNameShort[this.language][now.getDay()] :
+				BikramSambat.prototype.weekdayName[this.language][now.getDay()]);
+		this.labelDate.set_text(bsNow.getDateText());
+		this.labelMonthYear.set_text(bsNow.getMonthText()
+				+ (this.showYear !== "off" ? " " + (bsNow.getYearText().substring(this.showYear !== "full" ? 2 : 0)) : ""));
+		let time = NepaliPatro.zeroPad(now.getHours()) + ":"
+				+ NepaliPatro.zeroPad(now.getMinutes());
+		if (this.language !== "en")
+			time = BikramSambat.prototype.digitsEnToNe(time);
+		this.labelTime.set_text(time);
+		let todayText = now.getDate() + " " + NepaliPatro.MONTH_NAMES[this.language][now.getMonth()]
+				+ " " + now.getFullYear();
+		if (this.language !== "en")
+			todayText = BikramSambat.prototype.digitsEnToNe(todayText);
+		this.tooltipToday.set_text(todayText);
+
+		// Setup loop to update values
+		this.timeout = Mainloop.timeout_add_seconds(this.showTime ? 1 : 10, Lang.bind(this, this.updateValues));
```

Comment:
When I remove the desklet, .xsession-errors is flooded with:

```
(cinnamon:176953): Clutter-CRITICAL **: 00:03:37.096: clutter_text_get_text: assertion 'CLUTTER_IS_TEXT (self)' failed
```
and Cinnamon crashes.

I think this is because `updateValues` is running on an interval and continues to run afterwards, but now its calling methods on stale actors that are removed from the stage. To stop it from running afterwards, you could set a variable like `this.beingRemoved = true;` in `on_desklet_removed`. Then at the beginning of `updateValues`, make sure it checks if the desklet is being removed with something like:

```js
if (this.beingRemoved) {
      this.timeout = 0;
      return false;
    }
```

Reply:
Cinnamon seems to crash on settings change too. Can you reproduce?

Reply:
@jaszhix...
1.	Shouldn't:
```JavaScript
on_desklet_removed: function() {
	Mainloop.source_remove(this.timeout);
}
```
..remove the ```updateValues()``` from the loop?

2.	I havn't encountered any crashes on settings change and have been using the Desklet on two systems since a month.

Reply:
Returning true in a source handler means loop, false means source should be removed. The source_remove is ignored because you're calling that function recursively at the end of `updateValues`. [Related doc](http://devdocs.baznga.org/glib20~2.50.0/glib.sourcefunc).

I'm currently using git master, its possible its a breaking change with Cinnamon 3.8. Are you using 3.6?

Reply:
@jaszhix, I'm on ```Cinnamon 3.6.7```

Reply:
I was able to fix the settings change issue by changing onSettingChanged like so:

```js
onSettingChanged: function() {
		if (this.timeout) {
			Mainloop.source_remove(this.timeout);
		}
		this.updateCalendar();
	},
```

Reply:
@jaszhix, updated the ```onSettingChanged```.

Reply:
Can you remove nepalipatro@deeppradha from this PR?

Reply:
@jaszhix, I did delete the folder... I wasn't aware the files would still be available...
Should I delete this PR and create a new one?

Reply:
Did you commit the deletion? Check with `git status` locally from the repo root, then run `git rm -rf ./nepalipatro@deeppradha`, `git commit -m "Remove nepalipatro@deeppradha"`, and push the branch. If you go to Files changed -> Jump to... -> you still see it in the files list. 

I noticed you are working off of your master branch. I recommend having master only track upstream master, and creating PR branches from your master. That makes it easier to keep PRs in sync with the latest changes on master. [Good summary](https://www.digitalocean.com/community/tutorials/how-to-create-a-pull-request-on-github) of PR workflow. 

After that, I think this can be merged. Not seeing any problems with calendar@deeppradhan now.

Reply:
@jaszhix, I have removed ```nepalipatro@deeppradhan```.